src/Microsoft.DotNet.CodeFormatter.Analyzers/ExplicitVariableTypeFixer.cs

```diff
@@ -0,0 +1,114 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.DotNet.CodeFormatter.Analyzers
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public class ExplicitVariableTypeFixer : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(ExplicitVariableTypeAnalyzer.DiagnosticId);
+
+        public override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            CancellationToken cancellationToken = context.CancellationToken;
+            SyntaxNode root = await context.Document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            SemanticModel model = await context.Document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            Diagnostic diagnostic = context.Diagnostics.First();
+            TextSpan diagnosticSpan = diagnostic.Location.SourceSpan;
+
+            SyntaxNode typeSyntaxNode = null;
+            ILocalSymbol varSymbol = null;
+            switch (GetRuleType(diagnostic))
+            {
+                case RuleType.RuleVariableDeclaration:
+                    var variableDeclarationNode = root.
+                        FindNode(diagnosticSpan).
+                        FirstAncestorOrSelf<VariableDeclarationSyntax>();
+                    Debug.Assert(variableDeclarationNode != null);
+                    typeSyntaxNode = variableDeclarationNode.Type;
+                    // Implicit typed variables cannot have multiple declartors
+                    varSymbol = (ILocalSymbol)model.GetDeclaredSymbol(variableDeclarationNode.Variables.Single(), cancellationToken);
+                    break;
+
+                case RuleType.RuleForEachStatement:
+                    var forEachStatementNode = root.
+                        FindToken(diagnosticSpan.Start).
+                        Parent.
+                        FirstAncestorOrSelf<ForEachStatementSyntax>();
+                    Debug.Assert(forEachStatementNode != null);
+                    typeSyntaxNode = forEachStatementNode.Type;
+                    varSymbol = model.GetDeclaredSymbol(forEachStatementNode, cancellationToken);
+                    break;
+
+                case RuleType.None:
+                    break;
+            }
+            if (varSymbol != null)
+            {
+                DocumentEditor documentEditor = await DocumentEditor.CreateAsync(context.Document, cancellationToken);context.RegisterCodeFix(
```

Comment:
You can move the creation of the DocumentEditor into the ReplaceVarWithExplicittype method and call TypeExpression(..) in there. It's generally good practice to defer most of the work to inside the delegate for the codeaction because that only runs when the actual fix is invoked as opposed to any code outside that will have to run when we need to determine if there is any fix at all.